pkg/azurelustre/nodeserver.go

```diff
@@ -357,3 +514,359 @@ func (d *Driver) ensureMountPoint(target string) (bool, error) {
+func newLustreVolume(volumeID string, params map[string]string) (*lustreVolume, error) {
+	var mgsIPAddress, azureLustreName, subDir string
+	// Shouldn't attempt to delete anything unless sub-dir is actually specified
+	retainSubDir := true
+	subDirReplaceMap := map[string]string{}
+	// validate parameters (case-insensitive).
+	for k, v := range params {
+		switch strings.ToLower(k) {
+		case VolumeContextMGSIPAddress:
+			mgsIPAddress = v
+		case VolumeContextFSName:
+			azureLustreName = v
+		case VolumeContextSubDir:
+			subDir = v
+			subDir = strings.Trim(subDir, "/")
+
+			if len(subDir) == 0 {
+				return nil, status.Error(
+					codes.InvalidArgument,
+					"Context sub-dir must not be empty or root if provided",
+				)
+			}
+
+			if _, ok := params[VolumeContextRetainSubDir]; !ok {
+				return nil, status.Error(
+					codes.InvalidArgument,
+					"Context retain-sub-dir must be provided when sub-dir is provided",
+				)
+			}
+		case VolumeContextRetainSubDir:
+			retainSubDirString := strings.ToLower(v)
+			if retainSubDirString != "true" && retainSubDirString != "false" {
+				return nil, status.Error(
+					codes.InvalidArgument,
+					"Context retain-sub-dir value must be either true or false",
+				)
+			}
+
+			retainSubDir = retainSubDirString == "true"
+
+			if _, ok := params[VolumeContextSubDir]; !ok {
+				return nil, status.Error(
+					codes.InvalidArgument,
+					"Context sub-dir must be provided when retain-sub-dir is provided",
+				)
+			}
+		case podNameKey:
+			subDirReplaceMap[podNameMetadata] = v
+		case podNamespaceKey:
+			subDirReplaceMap[podNamespaceMetadata] = v
+		case podUIDKey:
+			subDirReplaceMap[podUIDMetadata] = v
+		case serviceAccountNameKey:
+			subDirReplaceMap[serviceAccountNameMetadata] = v
+		case pvcNamespaceKey:
+			subDirReplaceMap[pvcNamespaceMetadata] = v
+		case pvcNameKey:
+			subDirReplaceMap[pvcNameMetadata] = v
+		case pvNameKey:
+			subDirReplaceMap[pvNameMetadata] = v
+		}
+	}
+
+	if len(mgsIPAddress) == 0 {
+		return nil, status.Error(
+			codes.InvalidArgument,
+			"Context mgs-ip-address must be provided",
+		)
+	}
+
+	azureLustreName = strings.Trim(azureLustreName, "/")
+	if len(azureLustreName) == 0 {
+		return nil, status.Error(
+			codes.InvalidArgument,
+			"Context fs-name must be provided",
+		)
+	}
+
+	volName := ""
+
+	volFromID, err := getLustreVolFromID(volumeID)
+	if err != nil {
+		klog.Warningf("error parsing volume id '%v'", err)
+	} else {
+		volName = volFromID.name
+	}
+
+	vol := &lustreVolume{
+		name:            volName,
+		mgsIPAddress:    mgsIPAddress,
+		azureLustreName: azureLustreName,
+		subDir:          subDir,
+		retainSubDir:    retainSubDir,
+		id: fmt.Sprintf(
+			volumeIDTemplate,
+			volName,
+			azureLustreName,
+			mgsIPAddress,
+			subDir,
+			retainSubDir),
+	}
+
+	if volFromID != nil && *volFromID != *vol {
+		klog.Warningf("volume context does not match values in volume id for volume %q", volumeID)
+	}
+
+	return vol, nil
+}
```

Comment:
We spoke offline re: this approach vs NodeStage/UnstageVolume.  It was good and useful information that should probably end up in here with a couple commented-out decls of NodeStage/UnstageVolume.  That way no one tries to do that later.

Reply:
Done